Fix take_along_dim negative index handling (#146211)

KaaustaaubShankar opened the pull request:
Fixes: #146211

This PR fixes an issue with `torch.take_along_dim()` not correctly handling negative indices. Previously, using negative values in the `indices` tensor caused an out-of-bounds error. This update wraps indices correctly, matching Python-style indexing semantics.

### 🔧 Changes
- Modified `_take_along_dim_helper` to apply modulo logic for dimension-safe negative indexing.
- Added a unit test `test_take_along_dim_negative_indices` to `test/test_indexing.py` to assert correctness of negative indexing behavior.

### 🧪 Testing
```bash
pytest test/test_indexing.py -k test_take_along_dim_negative_indices
```

KaaustaaubShankar committed:
Fix take_along_dim negative index handling (#146211)

KaaustaaubShankar commented:
@pytorchbot label "release notes: cpp"

KaaustaaubShankar committed:
fixed clang format

albanD commented on test/test_indexing.py:
```diff
@@ -843,6 +843,13 @@ def test_step_assignment(self, device):
         self.assertEqual(v[0].tolist(), [0, 3, 0, 4])
         self.assertEqual(v[1:].sum(), 0)
 
+    def test_take_along_dim_negative_indices(self) -> None:
```
You can update the generic test for take_along_dim to test these values in https://github.com/pytorch/pytorch/blob/5f4c8e4c896f9a1d89e1bf9b3bd92b3e7db5c3b9/torch/testing/_internal/common_methods_invocations.py#L2988
This is better than a one-off test.

albanD commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -2740,7 +2743,12 @@ Tensor take_along_dim(
         _take_along_dim_helper(self, indices, opt_dim.value());
-    return self_broadcasted.gather(dim, indices_broadcasted);
+    auto dim_size = self_broadcasted.size(dim);
+    auto fixed_indices = at::where(
```
You already did this in the _take_along_dim_helper function above right?

KaaustaaubShankar replied:
yeah forgot to remove that. I fixed that up and still passes the check!

KaaustaaubShankar committed:
removed repeated code

KaaustaaubShankar committed:
removed old test and added it to common_methods

KaaustaaubShankar committed:
Fix take_along_dim to wrap negative indices correctly

KaaustaaubShankar commented:
@albanD @mruberry Did the changes. Had to change the OpInfo and the init.py for take_along_dim

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2986,6 +2986,7 @@ def error_inputs_logcumsumexp(op_info, device, **kwargs):
                          error_regex='Dimension out of range')
 
 def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad, **kwargs):
+    print(f"kwargs: {kwargs}")
```
Old print I guess

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -3002,6 +3003,14 @@ def sample_inputs_take_along_dim(op_info, device, dtype, requires_grad, **kwargs
     yield SampleInput(
         make_arg((S, S)), gather_variable((S, S // 2), 0, S, True, device=device))
 
+        # ✏️ Negative indices sample — guarded against python_ref
```
nit: too much indentation

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -19740,7 +19749,7 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
            # See https://github.com/pytorch/pytorch/pull/78358
            check_batched_forward_grad=False,
-           sample_inputs_func=sample_inputs_take_along_dim,
+           sample_inputs_func=partial(sample_inputs_take_along_dim),
```
spurious partial

albanD reviewed (commented):
Ho yes that looks good !
Let's make sure CI is all green as well

KaaustaaubShankar committed:
stylistic changes

- got rid of the print statement
- fixed comment indentation
- got rid of partial
- fixed clang formatting errors i saw in some of the job using 'lint runner -a'

KaaustaaubShankar commented:
@albanD alright i think i got the changes you requested and i also noticed some clang formatting errors in the jobs so i fixed that rl quick.

albanD reviewed (approved):
Still green!

albanD commented:
@pytorchbot merge
